projects/npm-tools/packages/npm-scripts/src/sass/build.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const fs = require('fs');
+const globby = require('globby');
+const r2 = require('liferay-r2');
+const path = require('path');
+const sass = require('sass');
+
+const log = require('../utils/log');
+
+const SASS_EXCLUDE = [
+	'**/_*.scss',
+	'**/build/**',
+	'**/classes/**',
+
+	// Copy from CSSBuilderArgs.java
+
+	'**/_diffs/**',
+	'**/.sass-cache*/**',
+	'**/.sass_cache_*/**',
+	'**/_sass_cache_*/**',
+	'**/_styled/**',
+	'**/_unstyled/**',
+	'**/css/aui/**',
+	'**/css/clay/**',
+	'**/tmp/**',
+];
+
+const getClayPaths = () => {
+	try {
+		return require('@clayui/css').includePaths;
+	} catch (e) {
+		return [];
+	}
+};
+
+const clayIncludePaths = getClayPaths();
+
+async function collectSassFiles(baseDir, excludes = []) {
+	const excludePaths = [
+		...SASS_EXCLUDE.map((exclude) => path.join(baseDir, exclude)),
+		...excludes.map((exclude) => path.join(baseDir, exclude)),
+	];
+
+	return await globby([
+		path.join(baseDir, '**/*.scss'),
+		...excludePaths.map((exclude) => `!${exclude}`),
+	]);
+}
+
+function buildSass(file, output) {
+	const newFileName = file
+		.split('/')
+		[file.split('/').length - 1].replace('.scss', '.css');
+
+	const outputFile = path.join(output, newFileName);
+
+	const result = sass.renderSync({
+		file,
+		includePaths: clayIncludePaths,
+		outFile: outputFile,
+		sourceMap: true,
+	});
+
+	return [outputFile, result.css, result.map];
+}
+
+function copyCss(filePath, newDirectory) {
+	if (!fs.existsSync(newDirectory)) {
+		fs.mkdirSync(newDirectory, {recursive: true});
+	}
+
+	fs.copyFileSync(filePath, path.join(newDirectory, path.basename(filePath)));
+}
+
+const SASS_DIR = '.sass-cache';
+
+async function main(baseDir, {excludes, outputDir = '', rtl = false}) {
+	const files = await collectSassFiles(baseDir, excludes);
+
+	if (!files.length) {
+		log(`BUILD CSS: No files found.`);
+
+		return;
+	}
+
+	for (let i = 0; i < files.length; i++) {
+		const file = files[i];
+
+		const baseBuildDirectory = path.join(
+			baseDir,
+			outputDir,
+			path.parse(file).dir.replace(baseDir, '')
+		);
+
+		const sassBuildDirectory = path.join(baseBuildDirectory, SASS_DIR);
+
+		copyCss(file, baseBuildDirectory);
+
+		const [outputFile, css, sourceMap] = buildSass(
+			file,
+			sassBuildDirectory
+		);
+
+		if (!fs.existsSync(sassBuildDirectory)) {
+			fs.mkdirSync(sassBuildDirectory, {recursive: true});
+		}
+
+		if (rtl) {
+			fs.writeFileSync(
+				outputFile.replace('.css', '_rtl.css'),
```

Comment:
@wincent what would be a case where `/\.css$/` would succeed but `.css` would fail? Are you thinking if the filepath was something like `some/base/dir.css/contains/main.css`?

Reply:
Yep, that's exactly the kind of case I was thinking about. We might say, "we don't have any crazy paths like that", but months or years from now we _could_ wind up with one and if that happens, stuff will break. This kind of unanticipated thing happens often enough to be annoying.

As a rule of thumb, you don't have to be paranoid about things that _might_ happen in the future, but in cases like this, where the cost of saying _exactly what you mean_ as opposed to _something approximately like what you mean_, then you should do the former; eg.

- **Exactly what you mean:** "path ends with `.css`" → use an anchored regex → file suffix gets transformed.
- **Something approximately like what you mean:** "path contains `.css` as a substring" → use a string → first instance of substring gets replaced, which is currently what you want, but not guaranteed to be in the future.

Reply:
Makes sense, just wanted to make sure I was thinking through it similarly, thanks!